src/metaswitch/clearwater/queue_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,253 @@
+                    queue_config = json.loads(self.default_value())
+                else:
+                    queue_config = json.loads(self._last_value)
+            else:
+                queue_config = json.loads(etcd_value)
+        except: # pragma: no cover
+            queue_config = json.loads(self.default_value())
+
+        self._fsm.fsm_update(queue_config)
+        etcd_updated_value = json.dumps(queue_config)
+
+        # If we have a new state, try and write it to etcd.
+        if etcd_updated_value != self._last_value:
+            _log.debug("Writing updated queue config to etcd")
+
+            if not self.write_to_etcd(etcd_updated_value):
+                self.fsm_loop(etcd_updated_value)
+
+    # Write the new cluster view to etcd. We may be expecting to create the key
+    # for the first time.
+    def write_to_etcd(self, queue_config):
+        _log.debug("Writing state {} into etcd with index {}"
+                   .format(queue_config, self._index))
+        needs_no_retry = True
+
+        try:
+            if self._index:
+                self._client.write(self.key(), queue_config, prevIndex=self._index)
+            else: # pragma: no cover
+                self._client.write(self.key(), queue_config, prevExist=False)
+        except (EtcdAlreadyExist, ValueError): # pragma: no cover
+            _log.debug("Contention on etcd write")
+            # Our etcd write failed because someone got there before us. We
+            # don't need to retry in this case as we'll just pick up the
+            # changes in the next etcd read
+        except Exception as e:
+            # Catch-all error handler (for invalid requests, timeouts, etc) -
+            # unset our state and start over.
+            _log.error("{} caught {!r} when trying to write {} with index {}"
+                       " - pause before retrying"
+                       .format(self._ip, e, queue_config, self._index))
+
+            # Setting last_cluster_view to None means that the next successful
+            # read from etcd will trigger the state machine, which will mean
+            # that any necessary work/state changes get retried.
+            self._last_value, self._last_index = None, None
+
+            # Sleep briefly to avoid hammering a failed server
+            self.pause()
+            needs_no_retry = False
+     
+        return needs_no_retry
+
+    def remove_from_queue(self, successful):
+        # Get and parse the current value
+        self._last_value, self._index = self.read_from_etcd(wait=False)
+        queue_config = json.loads(self._last_value)
+
+        # What is the current status?
+        queue_status = status_in_queue(queue_config[constants.JSON_QUEUED], self._id)
+        restart_status = status_in_queue(queue_config[constants.JSON_COMPLETED], self._id)
+        error_status = status_in_queue(queue_config[constants.JSON_ERRORED], self._id)
+
+        if len(queue_config[constants.JSON_QUEUED]) > 0 and \
```

Comment:
I think this is "delete myself if I'm the first entry", but I had to puzzle it out - a comment would be useful.